platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/TimeLimitedMatcher.java

```diff
@@ -0,0 +1,171 @@
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.platform.util;
+
+import java.time.Duration;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Customized and reduced-scope RegEx matcher class to protect against runaway searches.
+ *
+ * <p>Instances of this class, created through one of its two factory methods, will either succeed
+ * or timeout, throwing a RuntimeException - specifically {@link TimeoutException} - in the latter
+ * case. It is incumbent on the caller to handle the runtime exception.
+ *
+ * <p>Example usage:
+ *
+ * <pre>
+ *   try {
+ *     Pattern pattern = Pattern.compile("bc");
+ *     TimeLimitedMatcher matcher = TimeLimitedMatcher.create(pattern, "abc123");
+ *     if (matcher.matches()) {
+ *       // do something with the match
+ *     }
+ *   } catch (TimeLimitedMatcher.TimeoutException e) {
+ *     // handle exception
+ *   }
+ * </pre>
+ */
+public class TimeLimitedMatcher {
+  // Limit the lifetime of this sequence to five seconds by default.
+  private static final Duration DEFAULT_DURATION = Duration.ofSeconds(5L);
+
+  private final Matcher matcher;
+
+  private boolean matched;
+
+  /**
+   * Creates a new Time-limited matcher instance to search for the regular expression {@code
+   * pattern} against the {@code inputString}.
+   *
+   * <p>This method executes the search before returning, throwing a {@link TimeoutException} if the
+   * regex search runs longer than {@link #DEFAULT_DURATION}.
+   *
+   * @param pattern RegEx search pattern
+   * @param inputString String to match against.
+   * @return TimeLimitedMatcher instance to use for reporting out match details
+   */
+  public static TimeLimitedMatcher create(Pattern pattern, CharSequence inputString) {
+    return create(pattern, inputString, DEFAULT_DURATION);
+  }
+
+  /**
+   * Creates a new Time-limited matcher instance to search for the regular expression {@code
+   * pattern} against the {@code inputString}.
+   *
+   * <p>This method executes the search before returning, throwing a {@link TimeoutException} if the
+   * regex search runs longer than {@code duration}.
+   *
+   * @param pattern RegEx search pattern
+   * @param inputString String to match against.
+   * @param duration the length of time to continue processing the regex before failing
+   * @return TimeLimitedMatcher instance to use for reporting out match details
+   */
+  public static TimeLimitedMatcher create(
+      Pattern pattern, CharSequence inputString, Duration duration) {
+    Matcher matcher = pattern.matcher(new TLCharSequence(inputString, calculateEndTime(duration)));
+    return new TimeLimitedMatcher(matcher);
+  }
+
+  /**
+   * Returns the input subsequence matched by the previous match.
+   *
+   * @return The (possibly empty) subsequence matched by the previous match, in string form
+   * @see Matcher#group()
+   */
+  public String group() {
+    return matcher.group();
```

Comment:
❗️ Although you are caching the `matched` result at creation which ensures that the timeout is accounted for for the matching duration, the `group()` methods goes back into the matcher and will invoke the timed limited char sequence and possibly throw that timeout exception again since these may be called later.

My suggestion would be to make the timed limited char sequence an inner class of the TimedLimitedmatcher and disable the timeout limiting once `matches()` has returned.

Reply:
Actually, the `group` methods are not re-triggering the `charAt` methods; however, it might not be bad to restrict the timeout checks to `matches`.

Reply:
They are triggering the `subSequence()` followed by a `toString()` which means the timeout exception could still occur.

Reply:
They are calling `subSequence()`, but not recalling `charAt()` on those subs. I still like the idea of restricting the timeout operation to the duration of `matches()`, though.

Reply:
Ah! You throw the timeout exception only from within the `charAt()`